### PR TITLE
Merging 99 commits into main, including all object storage and VectorDB code

### PR DESCRIPTION
**Pull request by russfellows**
# Pull Request: mlp-storage v3.0.0 — Object Storage, Streaming Checkpoints, VDB Benchmark, KV Cache Extensions

**From:** `russfellows/mlc-storage:main`  
**Into:** `mlcommons/storage:main`  
**Date:** March 31, 2026  
**Author:** Russ Fellows \<russ.fellows@mlcommons.org\>

---

## Overview

This PR contributes a large body of work developed on the `russfellows` fork across
approximately Q1 2026. The changes fall into five major areas:

1. **Object storage backend support** — three pluggable S3 libraries for training and data generation
2. **Streaming checkpoint framework** — producer-consumer I/O pipeline with multi-library backend support
3. **DLIO benchmark submodule update** — v2.0.1-22 with Parquet readers and multi-library S3 support
4. **Vector database benchmark** (`vdb_benchmark/`) — new Milvus-based benchmark package
5. **KV Cache benchmark extensions** — `--num-gpus`, `--tensor-parallel`, `--io-trace-log`, and `--enable-latency-tracing` flags

All changes include the upstream merge of `mlcommons/storage:main` (commit `4be40e6`,
`mlcommons/DLIO_local_changes` v2.0.1-22) with all conflicts resolved.

---

### What was added

The benchmark harness now supports three interchangeable S3-compatible storage libraries
for training workloads and data generation:

| Library | Protocols | Concurrency model | Framework support |
|---------|-----------|-------------------|-------------------|
| **s3dlio** | S3, Azure, GCS, `file://`, direct I/O | Async Rust, up to 64 concurrent GETs per DataLoader worker | PyTorch + TensorFlow |
| **s3torchconnector** | AWS S3 only | Sequential per DataLoader worker | PyTorch only |
| **minio** | S3-compatible | Thread pool, up to 16 threads per worker | PyTorch + TensorFlow |

Library selection is a single config-file field (`storage_library: s3dlio`) or
command-line flag; no code changes are required to switch between libraries.

### New DLIO workload configs (`configs/dlio/workload/`)

27 new YAML workload configurations covering:

- **UNet3D H100** — MinIO, s3dlio, and s3torchconnector variants (train + datagen)
- **LLaMA-3 8B checkpoint** — MinIO, s3dlio, s3torchconnector (checkpoint read/write)
- **ResNet-50 / PyTorch** — local file backend, s3dlio, s3torchconnector, Azure
- **Multi-endpoint** — round-robin and MPI-parallel load balancing with s3dlio
- **Hybrid storage** — multi-tier (local + object) configuration
- **Zero-copy datagen** — dgen-rs (Rust) producer-consumer pool configs

### Measured performance (UNet3D H100, s3dlio + MinIO, ~23.5 GB dataset)

System: single node, ~1.2 GB/s measured network ceiling, 168 × 146.6 MB NPZ files.

| MPI Ranks | Warm epoch duration | I/O Throughput | Samples/s | Scaling vs NP=1 |
|:---------:|:-------------------:|:--------------:|:---------:|:---------------:|
| NP=1 | ~78 s | 332 ± 0.7 MB/s | 2.37 | 1.0× |
| NP=2 | ~43 s | 664 ± 3.2 MB/s | 4.75 | 2.0× |
| NP=4 | ~23 s | 1,720 ± 125 MB/s | 12.31 | 5.2× |

NP=4 saturated the measured ~1.2 GB/s network ceiling; the super-linear number
reflects prefetch and page-cache effects. Throughput figures use wall-clock epoch
duration (DLIO `Ending epoch N` log line) for consistency across all libraries.

### New test scripts (`tests/object-store/`)

Per-library shell-script test suites covering datagen, train, checkpoint, and cleanup,
for each of the three libraries:

- `dlio_{s3dlio,minio,s3torch}_{datagen,train,checkpoint,cleanup,cycle}.sh`
- `test_mlp_{s3dlio,minio,s3torch}.sh`
- `test_s3dlio_formats.py` / `test_s3dlio_formats.sh` — standalone integration test
  running a full put / list-verify / get cycle for every DLIO data format
  (npz, npy, hdf5, tfrecord, csv) against a real S3-compatible endpoint

### Security — boto3 ban

`mlpstorage/ban_boto3.py` raises an explicit `ImportError` if `boto3` is imported
anywhere in the benchmark harness. This enforces use of the approved multi-library
backends and prevents accidental credential exposure through the boto3 credential chain.

### New documentation

- `docs/Object_Storage.md` — architecture and getting-started guide
- `docs/Object_Storage_Library_Setup.md` — per-library install and credential setup
- `docs/Object_Storage_Test_Guide.md` — step-by-step test execution
- `docs/Object_Storage_Test_Results.md` — recorded test results (ongoing)
- `docs/STORAGE_LIBRARIES.md` — library comparison and selection guide
- `docs/MULTI_ENDPOINT_GUIDE.md` — multi-endpoint / load-balancing configuration
- `docs/PARQUET_FORMATS.md` — Parquet byte-range reader details
- `docs/QUICK_START.md` — single-page getting-started reference
- `tests/object-store/dlio_mpi_object_results.md` — MPI scaling results (UNet3D H100)
- `tests/object-store/S3library_review_21-Mar.md` — prefetch-fairness analysis across all three libraries
- `tests/object-store/Object_Perf_Results.md`, `tests/object-store/s3dlio_performance_analysis.md`

---

### What was added

A new `mlpstorage/checkpointing/` package implements a pluggable streaming checkpoint
pipeline using a producer-consumer architecture. The design separates checkpoint data
generation (producer) from storage I/O (consumer), enabling:

- **155× faster data generation**: the dgen-rs Rust backend produces synthetic checkpoint
  data at ~239 GB/s vs. NumPy's ~1.54 GB/s on the same hardware.
- **Overlap of compute and I/O**: the producer fills memory buffers while the consumer
  writes previous buffers to storage, reducing total checkpoint time.

### Package structure

```
mlpstorage/checkpointing/
├── streaming_checkpoint.py        # producer-consumer orchestrator
├── storage_readers/
│   ├── base.py                    # abstract base class
│   ├── file_reader.py             # local filesystem (buffered + fadvise)
│   ├── s3dlio_reader.py           # s3dlio (S3 / Azure / GCS / file / direct)
│   ├── s3torch_reader.py          # s3torchconnector
│   └── minio_reader.py            # MinIO Python SDK
└── storage_writers/
    ├── base.py
    ├── file_writer.py             # local filesystem (buffered + O_DIRECT)
    ├── s3dlio_writer.py
    ├── s3torch_writer.py
    └── minio_writer.py
```

### Checkpoint backends

Six backends in total — four object-storage and two local-filesystem:

| Backend | Storage | Notes |
|---------|---------|-------|
| `s3dlio` | S3 / Azure / GCS | Async Rust, multi-endpoint load balancing |
| `s3torchconnector` | AWS S3 | AWS-official PyTorch connector |
| `minio` | S3-compatible | MinIO native Python SDK |
| `file` | Local filesystem | Buffered I/O with `fadvise(MADV_SEQUENTIAL)` |
| `local_fs` | Local filesystem | Explicit fadvise readahead |
| `direct_fs` | Local filesystem | `O_DIRECT` — bypasses page cache |

### Representative checkpoint throughput (measured on test system)

| Backend | Write throughput |
|---------|:----------------:|
| s3dlio (`file://`) | 810 MB/s |
| minio | 686 MB/s |
| s3torch | 720 MB/s |

### New documentation and tests

- `docs/Streaming-Chkpt-Guide.md` — complete architecture guide, benchmark results, tuning
- `tests/checkpointing/test_streaming_backends.py` — pytest suite covering all 6 backends
- `tests/checkpointing/compare_methods.py`, `demo_checkpoint_methods.sh`

---

## 3. DLIO Benchmark Submodule Update

The `dlio_benchmark` submodule pointer has been updated to commit `4be40e6`
(`mlcommons/DLIO_local_changes`, v2.0.1-22), which includes:

- **Multi-library S3 readers**: `NPZReaderS3Iterable` with pluggable dispatch for
  s3dlio, s3torchconnector, and minio; equivalent readers for TFRecord, HDF5, CSV
- **Parquet support**: byte-range reader for columnar checkpoint/dataset formats
- **Streaming iterable dataset** support for large-scale training workloads

`.gitmodules` URL updated from `russfellows/dlio_benchmark` to the canonical
`https://github.com/mlcommons/DLIO_local_changes`.

---

## 4. Vector Database Benchmark (`vdb_benchmark/`)

A new self-contained benchmark package for Milvus vector databases, contributed
under `vdb_benchmark/`.

### Scope

| Component | Description |
|-----------|-------------|
| `vdbbench/load_vdb.py` | Dataset loading and index building |
| `vdbbench/simple_bench.py` | QPS / recall benchmark |
| `vdbbench/enhanced_bench.py` | Extended metrics: recall@k, GT comparison, statistical summaries |
| `vdbbench/collection_mgr.py` | Interactive Milvus collection manager |
| `vdbbench/compact_and_watch.py` | Compaction monitoring utility |
| `vdbbench/config_loader.py` | YAML config loading |
| `stacks/milvus/standalone/minio/` | Milvus + MinIO Docker Compose stack |
| `stacks/milvus/standalone/s3/` | Milvus + external S3 Docker Compose stack |

### Index configurations supplied

- 1M and 10M vector datasets
- HNSW and DiskANN index types
- 512-dimension AISAQ index configuration
- Vector normalization for inner-product searches

### Test suite

Full pytest suite under `vdb_benchmark/tests/` covering: database connection,
index management, collection loading, vector generation, config parsing,
compaction, and recall verification.

---

### New CLI arguments

| Flag | Default | Description |
|------|---------|-------------|
| `--num-gpus N` | 1 | Total GPUs in the deployment; scales the effective GPU VRAM tier (`num_gpus × gpu-mem-gb`) |
| `--tensor-parallel N` | 1 | Tensor-parallelism degree; must be ≤ `--num-gpus`; warns if not a power of 2 |
| `--io-trace-log <path>` | None | Activates I/O trace mode (see below) |
| `--enable-latency-tracing` | False | Enables per-operation latency recording via bpftrace *(merged from upstream)* |

`--num-gpus` and `--tensor-parallel` were developed in an earlier feature branch
and are restored and reconciled here alongside the upstream `--enable-latency-tracing`
flag; all four arguments coexist without conflict.

### I/O trace-log mode (`--io-trace-log`)

When `--io-trace-log <path>` is specified, the benchmark runs in **pure logical
trace mode**: the full LLM inference simulation (prefill, decode, multi-turn,
eviction, prefix caching) executes normally, but no real I/O is performed. Every
KV cache operation is written to a structured CSV:

```
Timestamp, Operation, Object_Size_Bytes, Tier, Key, Phase
```

- `Tier`: `Tier-0` (GPU VRAM), `Tier-1` (CPU RAM), `Tier-2` (NVMe/persistent)
- `Operation`: `Read` or `Write`
- `Phase`: `Prefill`, `Decode`, or `Evict`

Paths ending in `.zst` are written as a streaming zstd-compressed CSV (recommended
for runs longer than a few minutes; typical 10–20× compression ratio).

The trace decouples *workload definition* from *storage validation*: the CSV can
be replayed by `fio`, `sai3-bench`, or any other storage benchmark tool against
real hardware, independently of the Python simulation runtime.

**New module:** `kv_cache_benchmark/kv_cache/tracer.py` — `IOTracer` class.

### `validate_args()` — extended validation

`workload.py:validate_args()` now validates the new numeric fields:
- `num_gpus >= 1`
- `tensor_parallel >= 1`, must be ≤ `num_gpus`, warning if not a power of 2

### Test suite fixes

All three arg-fixture locations in `kv_cache_benchmark/tests/test_kv_cache.py`
have been updated to include the new fields with safe defaults
(`num_gpus=1`, `tensor_parallel=1`, `io_trace_log=None`, `enable_latency_tracing=False`,
plus `prefill_only`, `decode_only`, `validation_trace`, `use_burst_trace`, `burst_trace_path`).

**Test results after fixes:** 211 passed, 23 skipped, 0 failures.

### KV cache proposal document renamed

`kv_cache_benchmark/docs/MLperf v3 KV cache proposal.md` → `MLperf_v3_KV_cache_proposal.md`
(spaces replaced with underscores to avoid URL encoding issues).

---

### `.gitignore`

Added exclusions for S3 credential files (`.env`, `env-fast`), `__pycache__`,
`*.pyc`, build artifacts, and test output directories.

### `pyproject.toml`

- Version bumped to `3.0.0`
- `VERSION` is now derived from package metadata (`importlib.metadata`) rather than
  a hardcoded string
- `pyyaml` added as a runtime dependency
- Optional `[compression]` extra added (`zstandard`) for `.zst` trace output

### `setup_env.sh`

New environment bootstrap script: creates the `.venv`, installs all dependencies,
and validates the installation.

### Integration test suite

New `tests/integration/` directory with standalone (non-pytest) test and
benchmark scripts:

- `test_dlio_storage.py`, `test_storage_library.py` — end-to-end DLIO + library tests
- `test_multi_endpoint.py`, `test_multi_endpoint_integration.py` — multi-endpoint s3dlio
- `benchmark_s3dlio_read.py`, `benchmark_s3dlio_write.py` — raw GET/PUT benchmarks
- `test_ab_comparison.py`, `benchmark_read_comparison.py` — A/B comparison across libraries
- `test_zerocopy_direct.py` — O_DIRECT zero-copy path
- `test_dlio_mpi.py`, `test_mpi_basic.py` — MPI scaling tests

---

## Files Changed Summary

| Category | Added | Modified |
|----------|:-----:|:--------:|
| `configs/dlio/workload/` | 27 | 0 |
| `docs/` | 10 | 0 |
| `kv_cache_benchmark/` | 3 | 6 |
| `mlpstorage/checkpointing/` | 13 | 0 |
| `mlpstorage/` (other) | 1 | 3 |
| `patches/` | 4 | 0 |
| `tests/checkpointing/` | 3 | 0 |
| `tests/configs/` | 4 | 0 |
| `tests/integration/` | 19 | 0 |
| `tests/object-store/` | 33 | 0 |
| `tests/unit/` | 1 | 0 |
| `vdb_benchmark/` | 38 | 0 |
| Top-level (README, pyproject, gitignore, gitmodules) | 2 | 4 |
| **Total** | **158** | **13** |

---

## Testing Summary

| Test suite | Result |
|-----------|--------|
| `kv_cache_benchmark/tests/test_kv_cache.py` | **211 passed, 23 skipped, 0 failures** |
| Object-store format integration (`test_s3dlio_formats.py`) | Pass — all DLIO formats (npz, npy, hdf5, tfrecord, csv) |
| Checkpoint backend smoke tests | Pass — all 6 backends |
| MPI scaling (UNet3D H100, NP=1/2/4, s3dlio + MinIO) | Pass — near-linear scaling to network ceiling |

---

## Notes for Reviewers

- **boto3 ban** (`mlpstorage/ban_boto3.py`): the harness must not use boto3.
  This module enforces the constraint at import time. If a different enforcement
  mechanism is preferred, it is straightforward to swap.
- **dgen-rs dependency**: the 155× data-generation speedup depends on the `dgen-rs`
  Rust wheel. The benchmark falls back gracefully to NumPy if dgen-rs is not installed;
  a warning is emitted.
- **S3 credentials**: all integration tests load credentials from a `.env` file
  excluded from git via `.gitignore`. No credentials appear in any committed file.
- **Concurrency fairness note**: a prefetch-fairness analysis
  (`tests/object-store/S3library_review_21-Mar.md`) documents that s3torchconnector
  currently fetches one file at a time per DataLoader worker while s3dlio uses up to
  64 concurrent async GETs. A fair A/B comparison requires either raising
  s3torchconnector's batch size or lowering s3dlio's concurrency. This is left as
  follow-on work.
- **TLS**: a TLS certificate validation fix is included for MinIO environments using
  self-signed certificates (`ssl_verify=False` toggle in the reader/writer base class).

### CODE REVIEW

**Comment by idevasena** [comment]
@russfellows @FileSystemGuy I just updated the one commit with my github handle associated with personal email as opposed to org email and pushed to PR. Its showing me as the updates are in progress. I will try `recheck` after that and see.

**Comment by idevasena** [comment]
recheck

**Comment by russfellows** [comment]
All of the "kv_cache_benchmark" code  conflicts I BELIEVE should all take my changes.  I had conflicts as well, and manually merged them to resolve.  

I suppose we can review the remaining changes, but I think my merge is solid.  

That just leaves the pesky CLA issues.  @idevasena , you are good, as I amended the commits earlier to your ID which passes.  So, it is only me the bot hates.

**Comment by FileSystemGuy** [comment]
Resolves #209 
Resolves #271

**Comment by FileSystemGuy** [comment]
The CLA bot says that russfellows-25020967 hasn't signed the CLA.  I can use the god-mode bypass of putting you into the whitelist in the cla.yml file if we cannot find any other way to unblock this.

**Comment by russfellows** [comment]
God mode here we come...

**Review by FileSystemGuy — DISMISSED** [review]

**Review by idevasena — APPROVED** [review]

**Review by FileSystemGuy — APPROVED** [review]

**Comment by russfellows** [comment]
Yeah!
